### PR TITLE
Add `spmd_axis_name` to vmap to allow constraining mapped PartitionSpecs.

### PR DESCRIPTION
**Pull request by pschuh**
This is added to allow `vmap(lambda x: with_sharding_constraint(x, None), spmd_axis_name = 'x')(x)` to lower to `with_sharding_constraint(x, P('x'))`. Previously, there was no way to specify explicitly what axis to apply to the `with_sharding_constraint`. This patch adds `spmd_axis_name` to BatchTrace so that the `with_sharding_constraint` batching rule can optionally constrain based on the provided axis.

### CODE REVIEW

**Review by mattjj — COMMENTED** [review]
Seeing `spmd_axis_name` being plumbed through next to `axis_name` everywhere (both parts of the trace-global-data 'payload' on the MainTrace), I wonder if we could plumb this more generically. By 'generically' I mean like the `ctx` which now get plumbed into all our translation rules: it's both cleaner (fewer explicit arguments everywhere) and more future-proof (in case we need to add or remove some plumbed info) than keeping these things splatted into explicit arguments.

Maybe `axis_name`, `axis_size`, and `spmd_axis_name` could all be plumbed into the `payload`, and the payload could be plumbed through e.g. the rules/utilities for higher-order primitives parameterized by jaxprs? I'm not sure.

WDYT?

(If this is urgent to land, we could choose to land-and-iterate.)

**Comment by pschuh** [comment]
How general would this be? Is it just a type for the batcher transform? In that case, i'm not so sure how useful that would be. This is slightly urgent though. People are training with patches.

**Comment by mattjj** [comment]
Hm I'm not sure I understand, but the utility of what I'm proposing is avoiding having to plumb multiple explicit arguments through all these helper functions. That is, the usefulness is immediately realized in the JAX codebase, not hypothetical! I wouldn't mind landing that as a cleanup though.

File: jax/_src/api.py [hunk 1]
```diff
@@ -1441,6 +1442,8 @@ def vmap(fun: F,
       axis so that parallel collectives can be applied.
     axis_size: Optional, an integer indicating the size of the axis to be
       mapped. If not provided, the mapped axis size is inferred from arguments.
+    spmd_axis_name: Optional, a hashable Python object to insert into any
+      mapped PartitionSpecs.
```
Comment (mattjj) [inline jax/_src/api.py:1446]: Can we remove the docs for now, or explicitly say this is an experimental feature and may change?

**Review by mattjj — APPROVED** [review]
LGTM to land and iterate.

**Pull request merged**